wa/workloads/recentfling/__init__.py

```diff
@@ -0,0 +1,130 @@
+    Tests UI jank on android devices.
+
+    For this workload to work, ``recentfling.sh`` and ``defs.sh`` must be placed
+    in ``~/.workload_automation/dependencies/recentfling/``. These can be found
+    in the `AOSP Git repository <https://android.googlesource.com/platform/system/extras/+/master/tests/workloads>`_.
+
+    To change the apps that are opened at the start of the workload you will need
+    to modify the ``defs.sh`` file. You will need to add your app to ``dfltAppList``
+    and then add a variable called ``{app_name}Activity`` with the name of the
+    activity to launch (where ``{add_name}`` is the name you put into ``dfltAppList``).
+
+    You can get a list of activities available on your device by running
+    ``adb shell pm list packages -f``
+    """
+    supported_platforms = ['android']
+
+    parameters = [
+        Parameter('loops', kind=int, default=3,
+                  description="The number of test iterations."),
+        Parameter('start_apps', kind=bool, default=True,
+                  description="""
+                  If set to ``False``,no apps will be started before flinging
+                  through the recent apps list (in which the assumption is
+                  there are already recently started apps in the list.
+                  """),
+        Parameter('device_name', kind=str, default=None,
+                  description="""
+                  If set, recentfling will use the fling parameters for this
+                  device instead of automatically guessing the device.  This can
+                  also be used if the device is not supported by recentfling,
+                  but its screensize is similar to that of one that is supported.
+
+                  For possible values, check your recentfling.sh.  At the time
+                  of writing, valid values are: 'shamu', 'hammerhead', 'angler',
+                  'ariel', 'mtp8996', 'bullhead' or 'volantis'.
+                  """),
+    ]
+
+    def initialize(self, context):  # pylint: disable=no-self-use
+        if self.target.get_sdk_version() < 23:
+            raise WorkloadError("This workload relies on ``dumpsys gfxinfo`` \
+                                 only present in Android M and onwards")
+
+        args = '-i {} '.format(self.loops)
+        if not self.start_apps:
+            args += '-N '
+        if self.device_name is not None:
+            args += '-d {}'.format(self.device_name)
+
+        self.cmd = "echo $$>{workdir}/pidfile; cd {bindir}; exec ./recentfling.sh {args}; rm {workdir}/pidfile".format(
+            workdir=self.target.working_directory,
+            bindir=self.target.executables_directory, args=args)
+        self.defs_host = context.resolver.get(File(self, "defs.sh"))
+        self.defs_target = self.target.install(self.defs_host)
+        self.recentfling_host = context.resolver.get(File(self, "recentfling.sh"))
+        self.recentfling_target = self.target.install(self.recentfling_host)
```

Comment:
I might suggest splitting the discovery of assets and installation into a separate method which is decorated with the ``@once`` decorator and called here instead?
This is to prevent the scripts being repeatedly deployed for multiple iterations. (Note: in that case these attributes would need to be set as class attributes instead.)

Reply:
What if I decorate `initialize()` and `finalize()` with `@once` instead?  Wouldn't that achieve the same?

Reply:
You are also doing your args manipulation and cmd generation in ``initialize`` so if you decorated the entire method you would not be able to have multiple configurations of the workload. As you're not doing anything else in your ``finalize`` method this should be fine to decorate.

Reply:
True, we are doing arg manipulation in `initialize()`.  @setrofim suggested putting it in `setup()`, but I decided to move it to `initialize()`.  Now I see why I was wrong.  I'll leave the arg manipulation in `setup()` and let `initialize()` be a method that only needs to be called once.